internal/networkservice/chains/sriovns/server.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/endpoint"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/clienturl"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/connect"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/mechanisms"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/mechanisms/recvfd"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/adapters"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/chain"
+	"github.com/networkservicemesh/sdk/pkg/tools/addressof"
+	"github.com/networkservicemesh/sdk/pkg/tools/token"
+
+	"github.com/networkservicemesh/cmd-forwarder-sriov/internal/networkservice/common/noop"
+	"github.com/networkservicemesh/cmd-forwarder-sriov/internal/networkservice/common/resetmechanism"
+)
+
+type sriovServer struct {
+	endpoint.Endpoint
+}
+
+// NewServer - returns an Endpoint implementing the SR-IOV Forwarder networks service
+//             - name - name of the Forwarder
+//             - authzServer - policy for allowing or rejecting requests
+//             - tokenGenerator - token.GeneratorFunc - generates tokens for use in Path
+//             - pciPool - provides PCI functions
+//             - resourcePool - provides SR-IOV resources
+//             - sriovConfig - SR-IOV PCI functions config
+//             - vfioDir - host /dev/vfio directory mount location
+//             - cgroupBaseDir - host /sys/fs/cgroup/devices directory mount location
+//             - clientUrl - *url.URL for the talking to the NSMgr
+//             - ...clientDialOptions - dialOptions for dialing the NSMgr
+func NewServer(
+	ctx context.Context,
+	name string,
+	authzServer networkservice.NetworkServiceServer,
+	tokenGenerator token.GeneratorFunc,
+	pciPool resourcepool.PCIPool,
+	resourcePool resourcepool.ResourcePool,
+	sriovConfig *sriovconfig.Config,
+	vfioDir, cgroupBaseDir string,
+	clientURL *url.URL,
+	clientDialOptions ...grpc.DialOption,
+) endpoint.Endpoint {
+	rv := sriovServer{}
+
+	connectChainFactory := func(class string) networkservice.NetworkServiceServer {
+		return chain.NewNetworkServiceServer(
+			clienturl.NewServer(clientURL),
+			connect.NewServer(ctx,
+				client.NewCrossConnectClientFactory(
+					name,
+					// What to call onHeal
+					addressof.NetworkServiceClient(adapters.NewServerToClient(rv)),
+					tokenGenerator,
+					noop.NewClient(class),
+				),
+				clientDialOptions...,
+			),
+		)
+	}
+
+	sriovChain := sriovChain(
+		pciPool,
+		resourcePool,
+		sriovConfig,
+		vfioDir, cgroupBaseDir,
+		connectChainFactory,
+	)
+
+	rv.Endpoint = endpoint.NewServer(
+		ctx,
+		name,
+		authzServer,
+		tokenGenerator,
+		mechanisms.NewServer(map[string]networkservice.NetworkServiceServer{
+			kernel.MECHANISM:   sriovChain,
+			vfiomech.MECHANISM: sriovChain,
+			noopmech.MECHANISM: connectChainFactory(cls.LOCAL),
+		}),
+	)
+
+	return rv
+}
+
+func sriovChain(
```

Comment:
I would prefer to have SR-IOV chain inside NewServer function, it is not so big.

Reply:
Fixed in another PR: https://github.com/networkservicemesh/sdk-sriov/pull/73
https://github.com/networkservicemesh/sdk-sriov/blob/54f053b5711978a72c56424a7aec60c08f54ac54/pkg/networkservice/chains/sriovns/server.go#L105